### ISSUE TITLE
on/off yesnoradio styling?

### ISSUE THREAD

**Issue by jools-r** [textpattern/textpattern#1235]
This is more a suggestion than an issue.

For years we've had two no / yes radio buttons in the preferences panels. It works fine, we know it well so there's no _need_ to change it. All the same, it's fairly basic. I had a go at a simple CSS-only approach to styling them as a yes/no widget that looks like this:

![yesnoradio_styling](https://user-images.githubusercontent.com/1203899/38460228-fa293bda-3ab5-11e8-82e8-dfe0262a6661.png)

Is that maybe of interest? It's very simple. No sliding animation or fancy but unclear buttons. It works with the existing markup and yesnoradio function so nothing else other than css needs changing.

This is the CSS. It looks like a lot, but it could be much simplified by:
- Adding a `yesnoradio` class to the containing `.txp-form-field` wrapper so that it can be targeted more specifically and the input type can be assumed as being a pair of radio buttons with labels.
- Removing empty spaces between the inputs and labels. That avoids the need for the `font-size: 0;` and `font-size: 13px;` trick.

```
.txp-tabs-vertical-group .txp-form-field-value {
    font-size: 0;
}
.txp-tabs-vertical-group .txp-form-field-value * {
    font-size: 13px;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio] {
    position: absolute !important;
    clip: rect(0, 0, 0, 0);
    height: 1px;
    width: 1px;
    border: 0;
    overflow: hidden;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio] + label {
  display: inline-block;
  width: 3em;
  background-color: none;
  color: rgba(0, 0, 0, 0.3);
  font-size: 13px;
  font-weight: normal;
  text-align: center;
  text-shadow: none;
  padding: 2px 2px;
  border: 1px solid #ccc;
  transition: all 0.1s ease-in-out;
  margin-right: 0;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio] + label:hover {
    cursor: pointer;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio]:checked + label {
  background-color: #F4F4F4;
  box-shadow: none;
  color: #000;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio] + label:first-of-type {
  border-radius: 3px 0 0 3px;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio] + label:last-of-type {
  border-radius: 0 3px 3px 0;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio]:not([checked]) + label:first-of-type {
  border-right: 0;
}
.txp-tabs-vertical-group .txp-form-field-value input[type=radio]:not([checked]) + label:last-of-type {
  border-left: 0;
}
.txp-tabs-vertical-group .txp-form-field-value:hover input[type="radio"] + label {
    border-color: #ababab;
}
.txp-tabs-vertical-group .txp-form-field-value:hover input[type=radio]:checked + label:last-of-type {
    border-left-color: #ccc;
}
.txp-tabs-vertical-group .txp-form-field-value:hover input[type=radio]:checked + label:first-of-type {
    border-right-color: #ccc;
}
```

If you want to try it out, stick it into a file called `custom.css` in your `admin_themes/hive/assets/css` folder. No changes are needed to the yesnoradio function.

I've tried it on current firefox, chrome and safari but not IE / Edge or older versions. With German it also worked fine.

**Comment by Bloke** [comment]
I like it. I think Phil tried something like this a while ago (using jQuery UI I think?) but it was either r-e-a-l-l-y slow or required heavy code or something went amok in some browsers so we ditched it and went back to radios.

Things have moved on since then. Maybe we can look at this again, I'll leave it to Phil to decide. I'm fine with either look.

**Comment by phiw13** [comment]
Yeah, there is no need for JQ-UI to do this kind of things, although I’m not sure how accessible it is.

A few notes:

> * Adding a yesnoradio class to the containing .txp-form-field wrapper so that it can be targeted more specifically and the input type can be assumed as being a pair of radio buttons with labels.
> * Removing empty spaces between the inputs and labels. That avoids the need for the font-size: 0; and font-size: 13px; trick.

Adding a class to `txp-form-field` would surely simplify the code (suggestion: `txp-form-field-yesnoradio` ?). That would make it possible to use `display: flex`, there is no need to to remove the white-space between the inputs and labels.

You need to add some styling for the focussed state: `[type=radio]:focus + label`.

Demo: https://dev.l-c-n.com/txp/_issues/txp-radio.html

Need for some fudging for RTL languages

**Comment by philwareham** [comment]
Hi. I’ll add in some class names if you wish but I’m fairly happy with the current radios for now in 4.7.

Maybe we can look again at 4.8. I’m away until Tuesday though so I can’t add any class names until then.

**Comment by jools-r** [comment]
@philwareham: No problem, have a nice weekend! The added html-type class to the container might be useful anyway. And perhaps with Philippe's improvements, it's an easy gain?

@phiw13: I like your demo. Did you already have that in Sandspace? If so, I'm sorry I hadn't noticed it before.
Also wondered how well that works from an ARIA standpoint (my knowledge is still a bit basic on that front). Please improve on it by all means. It really was just a quick play on my part.

**Comment by phiw13** [comment]
> I like your demo. Did you already have that in Sandspace? If so, I'm sorry I hadn't noticed it before.

I just recycled some style already available in Sandspace. If/when Phil adds some classes, I‘ll improve it a little further. I don’t think it will be part of the theme, but I’ll put it in the custom_example stylesheet as an option.

What I’m a little worried about is the hiding of the radio-buttons; I’m not sure screen-readers will “see” them as such (and the problem I always have is listening to VoiceOver - hearing problems). Keyboard users shouldn't have a problem, except _maybe_ understanding that it is a pair of radio-buttons (press left/right arrow to toggle the state).

**Comment by philwareham** [comment]
Yes, jQuery UI reconned the aria states and roles onto the labels I believe. CSS alone is not enough.

**Comment by jools-r** [comment]
> Keyboard users shouldn't have a problem

> Yes, jQuery UI retconned the aria states and roles onto the labels I believe. CSS alone is not enough.

Tabbing and left and right arrows still works as before and as the rest of the markup is unchanged so the aria states and roles (or lack of) are exactly as they currently are. The addition of a :focus state as Philippe suggested would make one's tabbing position apparent.

> What I’m a little worried about is the hiding of the radio-buttons

The radio inputs are hidden with `clip` and `overflow:hidden;` and **not** `display:none;` or `visibility:hidden;`, so they should be detectable as before via screen readers (though I have not checked that).

Running an accessibility analysis using [axe devtools](https://addons.mozilla.org/de/firefox/addon/axe-devtools/) on the page throws up a few notes, mostly minor. Where the yesnoradio items are concerned, it says:

> Radio inputs with the same name attribute value must be part of a group. Ensures related <input type="radio"> elements have a group and that the group designation is consistent.

In short: this is just a visual tweak.

**Comment by jools-r** [comment]
Granted it's not as compact from a CSS point of view, but adding the html-type as a class to the `.txp-form-field` container is easier than adding it to `.txp-form-field-value`:

Change [line 264](https://github.com/textpattern/textpattern/blob/dev/textpattern/include/txp_prefs.php#L264) of txp_prefs.php to:

```
'class' => 'txp-form-field '.str_replace('_', '-', $a['html']),
```

To add it to `.txp-form-field-value`, it looks like the [inputLabel](https://github.com/textpattern/textpattern/blob/dev/textpattern/lib/txplib_html.php#L861) function needs changing and the class needs to be passed to it.

**Comment by philwareham** [comment]
That assumes the on/off yes/no radio buttons are restricted to prefs only. That's not the case.

**Comment by jools-r** [comment]
True. Good point.

**Comment by petecooper** [comment]
How's this looking? Worth bumping to 4.7.1 or 4.8 so we can get the ball rolling on 4.7 release?

**Comment by Bloke** [comment]
Pants, forgot about this, sorry. Will see if I can stuff it in over the weekend. If not, we'll defer it.

**Comment by philwareham** [comment]
I looked at the code and couldn't find a way to add a a class easily, so I bumped it to Stef https://github.com/textpattern/textpattern/issues/1235#issuecomment-380032823. I could probably be flagged as a 4.7.1 thing though as I'm not going to directly style those checkboxes (rather give others the ability to if they wish in their themes).

**Comment by petecooper** [comment]
+1 for deferring and saving @Bloke some weekend work.

**Comment by philwareham** [comment]
OK, pushed t a 4.7.1 issue.

**Comment by philwareham** [comment]
Pushing this to 4.8, as it would be easier to do as part of new widget framework.

**Comment by phiw13** [comment]
A few notes:

1/ there is absolutely no need to use the jQuery-UI stuff (and bloat). Everything can be done natively. (modulo adding a class to the container element.)

2/ a collection of on/off toggles: https://codepen.io/collection/LDKps/

w
3/ some notes by Adrian Roselli: http://adrianroselli.com/2019/03/under-engineered-toggles.html

**Comment by philwareham** [comment]
@phiw13 nice find, I'll read the Adrian Roselli blog post and test out some of the code when I get some spare time.

**Comment by jools-r** [comment]
Thanks Philippe and Phil! I agree, jQueryUI is overkill. Most of the demos I found when I looked around were for toggles based on checkboxes whereas here we have two radio buttons.

I use my low-tech variant for now and Philippe's already made some suggestions on how to improve on it. It's not as obvious or animated as some of those demos but I was trying to keep to your styling...

**Comment by philwareham** [comment]
@bloke I think this can be part of the new widget UI (and bumped to 4.9 release as discussed earlier today).
